content/articles/building-a-custom-stopwatch-npm-library-in-typescript/index.md

```diff
@@ -0,0 +1,340 @@
+
+In the `stopwatch.ts` file, implement the code snippet below:
+
+```typescript
+export function startWatch(timerStop: boolean, watchcycle: any) {
+  if (timerStop) {
+    timerStop = false;
+    watchcycle();
+  }
+}
+
+export function stopWatch(timerStop: boolean) {
+  if (timerStop == false) {
+    timerStop = true;
+  }
+}
+
+export function watchcycle(
+  timerStop: boolean,
+  milliseconds: number,
+  seconds: number,
+  minutes: number,
+  hours: number,
+  timer: HTMLElement
+): number | string {
+  if (timerStop == false) {
+    milliseconds = milliseconds + 10;
+
+    if (milliseconds == 1000) {
+      milliseconds = 0;
+      seconds = seconds + 1;
+    }
+    if (seconds == 60) {
+      minutes = minutes + 1;
+      seconds = 0;
+    }
+    if (minutes == 60) {
+      hours = hours + 1;
+      minutes = 0;
+      seconds = 0;
+    }
+
+    let hr = hours < 10 ? "0" + hours : hours;
+    let mins = minutes < 10 ? "0" + minutes : minutes;
+    let secs = seconds < 10 ? "0" + seconds : seconds;
+    let milsec =
+      milliseconds < 10
+        ? "00" + milliseconds
+        : milliseconds < 100
+        ? "0" + milliseconds
+        : milliseconds;
+
+    setTimeout("watchcycle()", 1000);
+    return (timer.innerHTML = hr + ":" + mins + ":" + secs + ":" + milsec);
+  }
+}
+
+export function resetWatch(
+  milliseconds: number,
+  seconds: number,
+  minutes: number,
+  hours: number,
+  timer: HTMLElement
+) {
+  hours = 0;
+  minutes = 0;
+  seconds = 0;
+  milliseconds = 0;
+  return (timer.innerHTML = "00:00:00");
+}
+```
+
+As you are already familiar with Typescript, we are expressly assigning a value type to every parameter each function is accepting (i.e. number, string, HTMLElement, and any).
+
+From the snippet above, we created and exported 4 functions as explained below:
+
+- `Startwatch`: This function is called when the `start` button is clicked. On execution, the function sets the `timerstop` to false and also starts the `watchcycle` function.
+- `Stopwatch`: This function runs when the user clicks the `stop` button. When executed, the `stopwatch` function sets the `timerstop` to true which halts the `watchcycle` function.
+- `resetwatch`: The `resetwatch` function simply returns the stopwatch to the initial value (0) and it is executed when the user clicks the `reset` button.
+- `watchcycle`: This is the brain behind our stopwatch, it increments the values of the milliseconds, seconds, minutes, and hours accordingly and then returns the values to the `timer` HTMLElement element tag in the `index.html` when used.
+
+That is all we need to set up the stopwatch logic. Finally, create an `index.ts` in `src` and export all the functions as shown below:
+
+```typescript
+export { startWatch, watchcycle, stopWatch, resetWatch } from "./stopwatch";
+```
+
+### Step 6 – Compiling the Typescript into JavaScript and publishing to NPM

```

Comment:
Publishing the package to NPM comes before testing it. This is not a good idea. Please move this step to be the last step.

Reply:
we are not testing the accuracy of the package per se. we are simply consuming the published package locally. I have updated the title to `consuming the package locally`, hope it's okay.

Reply:
Hello @davekola 

As a reader, it would be appropriate to know what you are publishing to NPM. At this point the reader has not utilized the package. The publishing step makes more sense to be last step after the reader has fully interacted with the package.

Reply:
@geoffrey45 understood
i have updated the article to accommodate your suggestions. thanks